porch/pkg/engine/engine.go

```diff
@@ -440,61 +442,93 @@ func (cad *cadEngine) ListFunctions(ctx context.Context, repositoryObj *configap
 	}
 
-	var ref string
-	var packageDir string
-	upstream := m.task.Clone.Upstream
-	switch upstream.Type {
-	case api.RepositoryTypeGit:
-		ref = upstream.Git.Ref
-		packageName := filepath.Base(m.task.Clone.Upstream.Git.Directory)
-		packageName = strings.TrimPrefix(packageName, ".git")
-		packageDir = filepath.Join(dir, packageName)
-	case api.RepositoryTypeOCI:
-		return repository.PackageResources{}, nil, fmt.Errorf("update is not supported for oci packages")
-	default:
-		revision, err := (&PackageFetcher{
-			cad:               m.cad,
-			referenceResolver: m.referenceResolver,
-		}).FetchRevision(ctx, upstream.UpstreamRef, m.namespace)
-		if err != nil {
-			return repository.PackageResources{}, nil, fmt.Errorf("error fetching the resources for package revisions %s", upstream.UpstreamRef.Name)
-		}
-		revisionKey := revision.Key()
-		ref = fmt.Sprintf("%s/%s", revisionKey.Package, revisionKey.Revision)
-		packageDir = dir
+	originalResources, err := (&PackageFetcher{
+		cad:               m.cad,
+		referenceResolver: m.referenceResolver,
+	}).FetchResources(ctx, currUpstreamPkgRef, m.namespace)
+	if err != nil {
+		return repository.PackageResources{}, nil, fmt.Errorf("error fetching the resources for package %s with ref %+v",
+			m.pkgName, *currUpstreamPkgRef)
 	}
 
-	if err := kpt.PkgUpdate(ctx, ref, packageDir, kpt.PkgUpdateOpts{}); err != nil {
-		return repository.PackageResources{}, nil, err
+	upstreamRevision, err := (&PackageFetcher{
+		cad:               m.cad,
+		referenceResolver: m.referenceResolver,
+	}).FetchRevision(ctx, targetUpstream.UpstreamRef, m.namespace)
+	if err != nil {
+		return repository.PackageResources{}, nil, fmt.Errorf("error fetching revision for target upstream %s", targetUpstream.UpstreamRef.Name)
+	}
+	upstreamResources, err := upstreamRevision.GetResources(ctx)
+	if err != nil {
+		return repository.PackageResources{}, nil, fmt.Errorf("error fetching resources for target upstream %s", targetUpstream.UpstreamRef.Name)
 	}
 
-	loaded, err := loadResourcesFromDirectory(dir)
+	klog.Infof("performing pkg upgrade operation for pkg %s resource counts local[%d] original[%d] upstream[%d]",
+		m.pkgName, len(resources.Contents), len(originalResources.Spec.Resources), len(upstreamResources.Spec.Resources))
+
+	// May be have packageUpdater part of engine to make it easy for testing ?
+	updatedResources, err := (&defaultPackageUpdater{}).Update(ctx,
+		resources,
+		repository.PackageResources{
+			Contents: originalResources.Spec.Resources,
+		},
+		repository.PackageResources{
+			Contents: upstreamResources.Spec.Resources,
+		})
 	if err != nil {
-		return repository.PackageResources{}, nil, err
+		return repository.PackageResources{}, nil, fmt.Errorf("error updating the package to revision %s", targetUpstream.UpstreamRef.Name)
+	}
+
+	newUpstream, newUpstreamLock, err := upstreamRevision.GetUpstreamLock()
```

Comment:
Does this still work with the changes I made to `GetUpstreamLock` in https://github.com/GoogleContainerTools/kpt/pull/3301?

Reply:
Good point. This introduces a bug :). This won't work. I need the upstreamRevision's (revision and revisionLock info) instead of upstreamRevision's upstream here.  It exists in the code, need a way to plumb it through.

Reply:
Addressed this in the latest commit: https://github.com/GoogleContainerTools/kpt/pull/3289/commits/2f6b7bd384c78b1a0069c29c18d12bb6501af669 PTAL